# Kapitel 4 - Klassificering

## Faktafrågor

### 1. Vad kännetecknar klassificeringsproblem? Ge några exempel på tillämpningsområden.

Ett klassificeringsproblem handlar om att gissa vilken **kategori** något tillhör. Svaret är alltid en av ett begränsat antal möjliga klasser.

Exempel:

- Gissa om ett mejl är spam eller inte.
- Gissa vilken siffra (0-9) någon har skrivit för hand.
- Gissa om en bild föreställer en katt eller en hund.
- Gissa om en patient är sjuk eller frisk utifrån provsvar.
- Gissa om en anställd kommer sluta på jobbet eller inte.

### 2. Förklara hur OvO och OvR-algoritmerna fungerar.

Vissa modeller (t.ex. SVM) kan från början bara skilja mellan **två** klasser åt gången. OvO och OvR är två sätt att ändå använda sådana modeller när man har **fler än två** klasser.

- **OvR (One-vs-Rest)** – man tränar en modell per klass, som svarar på frågan "är det här klass A, eller något annat?". Har man 3 klasser tränar man alltså 3 modeller. Den modell som är mest säker på sin förutsägelse vinner.

- **OvO (One-vs-One)** – man tränar en modell för **varje par** av klasser, t.ex. "A mot B", "A mot C", "B mot C". Har man 3 klasser blir det 3 modeller (fler klasser ger snabbt väldigt många modeller). Den klass som vinner flest "dueller" mot de andra klasserna blir slutgiltig gissning.

### 3. Förklara följande utvärderingsmått

a) **Confusion matrix** – en tabell som visar hur många gissningar som blev rätt eller fel, uppdelat per klass. Raderna visar det sanna svaret, kolumnerna visar vad modellen gissade. Siffrorna på diagonalen är de rätta gissningarna.

b) **Accuracy** – hur stor andel av alla gissningar som var korrekta, totalt sett.

c) **Precision** – av alla gånger modellen gissade "positiv" (t.ex. "spam"), hur stor andel var faktiskt positiva? Mäter hur pålitlig en positiv gissning är.

d) **Recall** – av alla observationer som faktiskt VAR positiva, hur stor andel lyckades modellen hitta? Mäter hur bra modellen är på att inte missa positiva fall.

e) **F1-score** – ett mått som väger ihop precision och recall till ett enda tal, användbart när man vill ha balans mellan de två istället för att bara titta på ett av dem.

f) **ROC-kurvan** – en graf som visar hur bra modellen är på att skilja mellan klasserna vid olika tröskelvärden. Ju mer kurvan böjer sig mot övre vänstra hörnet, desto bättre modell.

### 4. Vad är *precision-recall tradeoff* för något?

Precision och recall drar ofta åt olika håll: höjer du modellens "säkerhetskrav" (tröskelvärde) innan den gissar "positiv", blir gissningarna mer pålitliga (högre precision), men modellen missar fler faktiska positiva fall (lägre recall). Sänker du kravet istället hittar du fler positiva fall (högre recall), men fler av gissningarna blir felaktiga (lägre precision).

Man kan alltså sällan maximera båda samtidigt, man får välja en balans utifrån vad som är viktigast i just det problemet.

### 5. Förklara översiktligt hur respektive modell fungerar.

a) **Logistisk regression** – trots namnet en klassificeringsmodell. Räknar ut en sannolikhet (mellan 0 och 1) för att något tillhör en viss klass, med hjälp av en s-formad kurva, och klassar sedan utifrån ett tröskelvärde (t.ex. över 50% = klass 1).

b) **Support vector machines (SVM)** – försöker hitta den bästa möjliga gränsen (linje/plan) som skiljer klasserna åt, med så stor marginal som möjligt till närmaste datapunkter.

c) **Beslutsträd** – ställer en serie ja/nej-frågor om datan tills man hamnar i en slutgrupp, och läser av vilken klass den gruppen tillhör.

d) **Ensemble learning** – kombinerar flera modeller och låter dem "rösta" fram ett gemensamt svar, i hopp om bättre resultat tillsammans.

e) **Random forest** – många beslutsträd som tillsammans röstar fram en klass. Blir oftast stabilare än ett enda träd.

f) **Extra trees** – liknar random forest, men väljer även ett **slumpmässigt** tröskelvärde vid varje förgrening istället för att leta upp det bästa. Går snabbare att träna, ger ofta lite högre bias men lägre variance.

### 6. Vad innebär det att vi kan kolla på *feature importance* med hjälp av trädmodeller såsom beslutsträd eller *random forest*? 

Trädmodeller kan visa hur mycket varje variabel faktiskt bidrog till att göra bra uppdelningar av datan. Man får ut en siffra per variabel – ju högre siffra, desto viktigare var den variabeln för modellens gissningar. Alla siffrorna summerar till 1.

Det är användbart av två skäl: dels för att förstå vilka variabler som faktiskt spelar roll, dels för att kunna välja bort variabler med väldigt låg vikt (variabelselektion), även om man tänker använda en helt annan modell för de faktiska prediktionerna.

## Resonemangfrågor


### 7. Stina säger "jag vill ha högsta möjliga precision". Kalle undrar vad som händer med recall då. Vad hade du svarat?

Stina har rätt i att om man höjer tröskeln för när modellen ska säga att något är positivt, ökar precisionen. Modellen blir då mer försiktig med att säga "positiv" och gör färre falskt positiva gissningar. Samtidigt sjunker ofta recall, eftersom modellen då missar fler av de faktiska positiva fallen.

**Hög precision är bra när** ett falskt positivt är kostsamt eller pinsamt, t.ex. ett spamfilter som av misstag flyttar viktiga mejl till skräpposten.

**Hög precision kan däremot innebära lägre recall**, vilket kan vara ett problem när det är farligt att missa faktiska positiva fall. Ett exempel är en modell som ska upptäcka en allvarlig sjukdom. Där kan det vara bättre att få några falska larm (lägre precision) än att missa en sjuk patient (lägre recall).

**I rättsväsendet:** en felaktig fällande dom kan få mycket allvarliga konsekvenser. I ett sådant fall kan man därför vilja minimera antalet falskt positiva fall, även om det innebär att modellen missar vissa positiva fall. Det brukar sammanfattas med principen "hellre fria tio skyldiga än fälla en oskyldig".


### 8. Förklara hur man kan tolka figur 4.8 på sidan 175.

Bilden visar hur en logistisk regressionsmodell delar in ett område i två klasser, baserat på två variabler ($x_1$ och $x_2$).

- **Prickarna** är faktiska observationer. Lila prickar tillhör klass 0, gula prickar tillhör klass 1.
- **Bakgrundsfärgen** visar vad modellen tror om varje punkt i hela området, inte bara där det finns riktiga observationer. Ju mer lila bakgrunden är, desto mer säker är modellen på att en punkt där tillhör klass 0 (sannolikhet nära 0). Ju mer gul, desto säkrare är den på klass 1 (sannolikhet nära 1).
- **De gröna/turkosa banden i mitten** är det osäkra området, där sannolikheten ligger runt 50/50 – modellen är där inte särskilt säker på vilken klass det handlar om.
- **Beslutsgränsen** är där bakgrunden byter från mest-lila till mest-gul (ungefär mitt i de gröna banden). Det är gränsen modellen faktiskt använder för att bestämma "klass 0" eller "klass 1": punkter på den ena sidan klassas som 0, punkter på den andra som 1.

Man ser också att gränsen inte är perfekt – en del lila prickar hamnar på den gula sidan och tvärtom. Det är helt normalt, eftersom klasserna i verkligheten inte alltid går att separera helt perfekt med en rak linje (vilket är vad logistisk regression ritar, eftersom det är en linjär modell).

### 9. På sidan 209 står det "på träningsdatan använder vi .fit_transform(), på valideringsdatan och testdatan använder vi endast .transform()." Förklara logiken bakom detta.

`.fit_transform()` gör två saker samtidigt: den **lär sig** parametrarna (t.ex. medelvärde och standardavvikelse för StandardScaler) utifrån datan, och **skalar om** datan utifrån dessa parametrar direkt.

`.transform()` gör bara det andra steget – skalar om datan, men utifrån parametrar som redan lärts in tidigare (utan att beräkna nya parametrar).

Anledningen till att vi bara använder `.fit_transform()` på träningsdatan är att vi vill att modellen bara ska "se" och lära sig av träningsdatan. Om vi skulle använda `.fit_transform()` även på validerings- eller testdatan, skulle skalningen påverkas av information i den datan också – vilket är en form av **data leakage** (information läcker in som modellen egentligen inte borde ha tillgång till). Det skulle ge en alldeles för positiv bild av hur bra modellen faktiskt presterar på helt ny, osedd data. Genom att bara använda `.transform()` på validering/test appliceras exakt samma skalning som lärdes in på träningsdatan, vilket ger en ärlig utvärdering.

## Koduppgifter

### 11. Förklara vad nedanstående kod gör och tolka resultatet från koden. 

`y_true` är de sanna klasserna, `y_pred` är vad modellen gissade. `classification_report()` jämför de två och skriver ut precision, recall och F1-score för varje klass, samt accuracy.


In [1]:
from sklearn.metrics import classification_report

y_true = [0, 1, 2, 2, 2]
y_pred = [0, 0, 2, 2, 1]
target_names = ['class 0', 'class 1', 'class 2']
print(classification_report(y_true, y_pred, target_names=target_names))

              precision    recall  f1-score   support

     class 0       0.50      1.00      0.67         1
     class 1       0.00      0.00      0.00         1
     class 2       1.00      0.67      0.80         3

    accuracy                           0.60         5
   macro avg       0.50      0.56      0.49         5
weighted avg       0.70      0.60      0.61         5



Modellen presterar bäst på class 2, där precisionen är 1.00 och recall 0.67. För class 0 hittar modellen alla rätta fall (recall 1.00), men hälften av gissningarna är fel (precision 0.50). På class 1 presterar modellen sämst eftersom den inte hittar något av de faktiska fallen. Totalt sett blir accuracy 0.60, alltså 3 av 5 gissningar rätt. Det visar att en samlad accuracy kan dölja att modellen presterar olika bra på olika klasser.